src/qat/ir/instructions.py

```diff
@@ -0,0 +1,1032 @@
+
+    @field_validator("var_type", mode="before")
+    @classmethod
+    def _validate_var_type(cls, var_type):
+        # Types can't be (de)serialized, so we use a validator to manually
+        # find the correct type.
+        if isinstance(var_type, str):
+            var_type = locate(var_type)
+        return var_type
+
+    @field_validator("value", mode="after")
+    @classmethod
+    def _validate_value_type(cls, value, val_info):
+        var_type = val_info.data["var_type"]
+        if var_type != None and not isinstance(value, var_type):
+            raise ValueError(f"Value provided has type {type(value)}: must be {var_type}")
+        return value
+
+
+class IndexAccessor(Variable):
+    """Used to access an array index on a particular variable."""
+
+    index: int
+
+    def __init__(self, name, index, **kwargs):
+        super().__init__(name)
+        self.index = index
+
+    def __repr__(self):
+        return f"{self.name}[{self.index}]"
+
+
+### Instructions
+class Instruction(BaseModel):
+    inst: Literal["Instruction"] = "Instruction"
+
+
+class Repeat(Instruction):
+    """
+    Global meta-instruction that applies to the entire list of instructions. Repeat
+    value of the current operations, also known as shots.
+    """
+
+    inst: Literal["Repeat"] = "Repeat"
+    repeat_count: Any
+    repetition_period: Any = None  # TODO: probably can only be a float, but not sure...
+
+    def __init__(self, repeat_count: int = None, repetition_period=None, **kwargs):
+        super().__init__(repeat_count=repeat_count, repetition_period=repetition_period)
+
+    def __repr__(self):
+        return f"repeat {self.repeat_count},{self.repetition_period}"
+
+
+class QuantumMetadata(Instruction):
+    inst: Literal["QuantumMetadata"] = "QuantumMetadata"
+
+
+class Assign(Instruction):
+    """
+    Assigns the variable 'x' the value 'y'. This can be performed as a part of running
+    on the QPU or by a post-processing pass.
+    """
+
+    # set as any for now... not sure what the restrictions would be here
+    inst: Literal["Assign"] = "Assign"
+    name: str
+    value: Any
+
+    def __init__(self, name, value, **kwargs):
+        super().__init__(name=name, value=value)
+
+    def __repr__(self):
+        return f"{self.name} = {str(self.value)}"
+
+    @field_validator("value", mode="before")
+    @classmethod
+    def _list_to_vars(cls, val):
+        if isinstance(val, list):
+            lst = []
+            for itm in val:
+                if isinstance(itm, dict):
+                    lst.append(Variable(**itm))
+                else:
+                    lst.append(itm)
+            return lst
+        return val
+
+
+class Return(Instruction):
+    """A statement defining what to return from a quantum execution."""
+
+    inst: Literal["Return"] = "Return"
+    variables: Optional[List[str]]
+
+    def __init__(self, variables: List[str] = None, **kwargs):
+        if variables is None:
```

Comment:
Is there any need for this checking when you have the validator?

Reply:
Or just set the default value of `variables` to []?